### DISCUSSION TITLE
How can I test the accuracy of the model for each class

### DISCUSSION THREAD

**Discussion by 6sz** [open-mmlab/mmclassification#1089] [Q&A] [answered]
I want to know how accurate is the model for each class, or how can I get the Confusion matrix.Looking forward to getting help

**Comment by mzr1996** [comment, marked as answer]
Hello, we haven't exposed the confusion matrix calculation API by now. But you can call it directly.
For example, get the classification score of every sample at first:
```shell
python tools/test.py your_config_file your_checkpoint --out result.pkl --out-items class_scores
```
And then, calculate the confusion matrix:
```python
>>> import mmcv
>>> from mmcls.datasets import build_dataset
>>> from mmcls.core.evaluation import calculate_confusion_matrix
>>> cfg = mmcv.Config.fromfile("your_config_file")
>>> dataset = build_dataset(cfg.data.test)
>>> pred = mmcv.load("./result.pkl")['class_scores']
>>> matrix = calculate_confusion_matrix(pred, dataset.get_gt_labels())
>>> print(matrix)
tensor([[47.,  0.,  0.,  ...,  0.,  0.,  0.],
        [ 0., 46.,  0.,  ...,  0.,  0.,  0.],
        [ 0.,  0., 38.,  ...,  0.,  0.,  0.],
        ...,
        [ 0.,  0.,  0.,  ..., 36.,  0.,  0.],
        [ 0.,  0.,  0.,  ...,  0., 24.,  0.],
        [ 0.,  0.,  0.,  ...,  0.,  0., 26.]])
>>> import matplotlib.pyplot as plt
>>> plt.imshow(matrix[:20, :20])   # Visualize the first twenty classes.
>>> plt.show()
```
![图片](https://user-images.githubusercontent.com/26739999/145745823-086acd50-20be-4bfa-b7ad-da2644467913.png)

**Reply by SunTaoTao-666** [reply]
matrix = calculate_confusion_matrix(result['class_scores'], dataset.get_gt_labels())
result会报错，未定义

**Reply by mzr1996** [reply]
> matrix = calculate_confusion_matrix(result['class_scores'], dataset.get_gt_labels()) result会报错，未定义

fixed

**Comment by 6sz** [comment]
Wow, thank you very much

**Comment by mxmdpc** [comment]
There has an easy method to get each class ap.
1、You can change the file 'mean_ap.py' -> def mAP
![image](https://user-images.githubusercontent.com/71645197/161279165-51487964-e4d7-4178-91bb-9fc8c1f9dc6b.png)

2、return ap in the file of test.py,then print the ap array,like             
print(f'\apple: {round(aps[0], 2)}')
print(f'banana : {round(aps[1], 2)}')
![image](https://user-images.githubusercontent.com/71645197/161279410-db681f4e-5fbe-48fc-8b7b-b9243339f4a1.png)

**Comment by vothaianh1997** [comment]
.

**Comment by Anshulky** [comment]
Hey! what approach should i implement to get confusion matrix for multilabel?
